trans/transducer.py

```diff
@@ -539,6 +539,99 @@ def continue_decoding():
         return Output(action_history, self.decode_encoded_output(input_, action_history),
                       log_p, None)
 
+    def roll_in(self, sample: utils.Sample, rollin: int) -> None:
+        """Runs the transducer for greedy decoding.
+
+        Args:
+           sample: A single sample from the training data.
+           rollin: tbd.
+
+        Returns:
+            None."""
+
+        # adjust initial decoder states if batch_size has changed
+        self.h0_c0 = 1  # batch size is always zero
+
+        # initialize state variables
+        alignment_history = torch.tensor([[0]], device=self.device)
+        action_history = torch.tensor([[[BEGIN_WORD]]],
+                                      device=self.device, dtype=torch.int)
+        input_length = torch.tensor([len(sample.input) + 1], device=self.device)
+        output = []
+        optimal_actions = [[BEGIN_WORD]]
+
+        # run encoder
+        bidirectional_emb = self.encoder_step(sample.encoded_input.unsqueeze(dim=0))
+
+        # compute feature embedding
+        feature_emb = self.feature_embedding(sample.encoded_features)
+
+        # initial cell state for decoder
+        decoder = self.h0_c0
+
+        stop = False
+        while not stop and action_history.size(2) <= MAX_ACTION_SEQ_LEN:
+            current_alignment = alignment_history[:, -1]
+            valid_actions_mask = self.valid_actions_lookup[:, input_length - current_alignment]
+
+            # run decoder
+            decoder_output, decoder = self.decoder_step(
+                bidirectional_emb, feature_emb, decoder,
+                current_alignment, action_history[:, :, -1])
+
+            #  model's predictions
+            actions, log_probs = self.calculate_actions(decoder_output, valid_actions_mask)
+            log_probs_np = log_probs.squeeze().cpu().detach().numpy()
+            sample_action = self.sample(log_probs_np)
+
+            # expert prediction
+            expert_actions = self.expert_rollout(sample.input, sample.target,
+                                                 current_alignment.item(), output)
+            optimal_actions.append(expert_actions)
+
+            # update states
+            if np.random.rand() <= rollin:
```

Comment:
The next state is either sampled from the expert or the model itself. Is this the idea? Alternatively, one could always execute the model (and only rely on the expert for the targets).